src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingRedactionInterceptor.cs

```diff
@@ -0,0 +1,203 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if NET8_0_OR_GREATER
+
+using System;
+using System.Buffers;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.HttpLogging;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Compliance.Classification;
+using Microsoft.Extensions.Compliance.Redaction;
+using Microsoft.Extensions.Http.Diagnostics;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Diagnostics.Logging;
+
+internal sealed class HttpLoggingRedactionInterceptor : IHttpLoggingInterceptor
+{
+    private readonly IHttpLogEnricher[] _enrichers;
+    private readonly IncomingPathLoggingMode _requestPathLogMode;
+    private readonly HttpRouteParameterRedactionMode _parameterRedactionMode;
+    private readonly ILogger<HttpLoggingRedactionInterceptor> _logger;
+    private readonly IHttpRouteParser _httpRouteParser;
+    private readonly IHttpRouteFormatter _httpRouteFormatter;
+    private readonly IIncomingHttpRouteUtility _httpRouteUtility;
+    private readonly HeaderReader _requestHeadersReader;
+    private readonly HeaderReader _responseHeadersReader;
+    private readonly string[] _excludePathStartsWith;
+    private readonly FrozenDictionary<string, DataClassification> _parametersToRedactMap;
+
+    public HttpLoggingRedactionInterceptor(
+        IOptions<LoggingRedactionOptions> options,
+        ILogger<HttpLoggingRedactionInterceptor> logger,
+        IEnumerable<IHttpLogEnricher> httpLogEnrichers,
+        IHttpRouteParser httpRouteParser,
+        IHttpRouteFormatter httpRouteFormatter,
+        IRedactorProvider redactorProvider,
+        IIncomingHttpRouteUtility httpRouteUtility)
+    {
+        var optionsValue = options.Value;
+        _logger = logger;
+        _enrichers = httpLogEnrichers.ToArray();
+        _httpRouteParser = httpRouteParser;
+        _httpRouteFormatter = httpRouteFormatter;
+        _httpRouteUtility = httpRouteUtility;
+
+        _parametersToRedactMap = optionsValue.RouteParameterDataClasses.ToFrozenDictionary(StringComparer.Ordinal);
+
+        _requestPathLogMode = EnsureRequestPathLoggingModeIsValid(optionsValue.RequestPathLoggingMode);
+        _parameterRedactionMode = optionsValue.RequestPathParameterRedactionMode;
+
+        _requestHeadersReader = new(optionsValue.RequestHeadersDataClasses, redactorProvider);
+        _responseHeadersReader = new(optionsValue.ResponseHeadersDataClasses, redactorProvider);
+
+        _excludePathStartsWith = optionsValue.ExcludePathStartsWith.ToArray();
+    }
+
+    public ValueTask OnRequestAsync(HttpLoggingInterceptorContext logContext)
+    {
+        var context = logContext.HttpContext;
+        var request = context.Request;
+        if (ShouldExcludePath(context.Request.Path.Value!))
```

Comment:
Since this is a hot path, we can have a field like `isExclusionListEmpty` to not call the `ShouldExcludePath()` method if the `_excludePathStartsWith` is empty

Reply:
I'd expect the foreach to short circuit almost as fast without special handling.

Reply:
Method invocation will eat some CPU cycles